keps/168-2-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,313 @@
+- Data is consistent across all LocalQueues,
+- User with only LocalQueue visibility cannot access the list of pending workloads for ClusterQueue.
+
+### API Details
+
+We introduce a new API that will extend the existing one.
+
+There will be separate endpoints exposing the information about pending workloads for LocalQueues, and ClusterQueues. Each endpoint exposes information about a pending workload, such as:
+- workload's position in a ClusterQueue,
+- workload's position in a LocalQueue,
+- workload's priority,
+- creation timestamp.
+
+The API does not allow for the modification of any objects.
+
+Regular users will have access only to the LocalQueues they are assigned to. However, they will be able to fetch information about the global position of a workload in a ClusterQueue, without any details about workloads in different LocalQueues. 
+
+Administrators will have access to all the data at the ClusterQueue level. They will be able to view all the workloads, no matter the LocalQueues the workloads are assigned to.
+
+The API also allows user to fetch information about part of the Cluster/LocalQueue from position X to Y. There are two query parameters to do so:
+- `offset` indicates position of the first fetched workload - default: `0`
+- `limit` indicates max number of workloads to be fetched - default: `1000`
+
+Since the K8s Aggregation Layer does not support pagination out of the box, we do not want to commit to it since it would require significant effort. However, it can be implemented on the client side. We also provided above-mentioned query parameters that can help in cases where one does not want to fetch the whole Cluster/LocalQueue. Additionally, we expose [the endpoint to fetch information about a single pending workload](#fetch-information-about-a-single-pending-workload), so there is a way to query its position without listing all the remaining pending workloads.
+
+### API endpoints:
+
+We introduce a new API group ```visibility.kueue.x-k8s.io``` that aggregates following endpoints: 
+
+#### List all pending workloads in ClusterQueue
+
+```
+GET /apis/visibility.kueue.x-k8s.io/VERSION/clusterqueues/CQ_NAME/pending_workloads?offset=0&limit=1000
+```
+
+#### List all pending workloads in LocalQueue
+```
+GET /apis/visibility.kueue.x-k8s.io/VERSION/namespaces/LQ_NAMESPACE/localqueues/LQ_NAME/pending_workloads?offset=0&limit=1000
+```
+
+#### Fetch information about a single pending Workload
+```
+GET /apis/visibility.kueue.x-k8s.io/VERSION/namespaces/WL_NAMESPACE/pending_workload/WL_NAME
+```
+
+### API Objects:
+
+```
+// PendingWorkloadSummary is a user-facing representation of a pending workload in both LocalQueues and ClusterQueue that summarizes neccessary information from the admission order perspective
+type PendingWorkloadSummary struct {
+	TypeMeta TypeMeta
+	ObjectMeta ObjectMeta
+
+	LocalQueueName string
+	PositionInClusterQueue int32
+	PositionInLocalQueue int32
```

Comment:
Can we actually get this information? Otherwise, I wouldn't prioritize it.

Reply:
We don't have it at hand, but the computation is rather cheap - when we iterate over the snapshoted data for cluster queue, we increment position counters per local queue. Still, I agree it does not seem like a priority. I think we can do this later if needed for dashboards.